feat(docs): add uniswap tut (#20233)
AztecProtocol/aztec-packages: docs/docs-developers/docs/tutorials/js_tutorials/uniswap_swap.md

Reviewer: If the intention is to have readers write the code as they go along, I think it should be more clear. There should be a "project setup" section that helps readers set up their project so they know what tools they will be using and where to put the files as they encounter them.

---
title: "Cross-Chain Token Swap (L1 <> L2)"
sidebar_position: 2
description: "Build a cross-chain token swap that exits L2, swaps on L1, and deposits the output back to L2 using Aztec's messaging protocol."
references: ["docs/examples/contracts/example_uniswap/*", "docs/examples/solidity/example_swap/*", "docs/examples/ts/example_swap/*"]
---

## Why Build a Cross-Chain Swap?

DeFi liquidity lives on Ethereum L1. Users with tokens on Aztec L2 need a way to access L1 DEXs like Uniswap without manually bridging tokens back and forth. A cross-chain swap automates this: the user initiates the swap on L2, and the protocol handles exiting to L1, performing the swap, and depositing the output back to L2.

This tutorial walks you through building a version of this flow. You will learn how **L2-to-L1 messages** work and how multiple contracts across two chains coordinate to execute a single user action.

Before starting, make sure you have the Aztec local network running at version #include_aztec_version. Check out [the local network guide](../../../getting_started_on_local_network.md) for setup instructions.

> ## What You'll Build

Reviewer: It would be good to add a prerequisites section somewhere near the top. 

This quickly gets into a bit of Aztec jargon (l2-l1 messages, portals, governance concepts) and it would be nice for readers to know where to go to learn more about these topics. Maybe just add links in-line as these concepts are introduced

        TokenA["Token A (WETH)"]
        TokenB["Token B (DAI)"]
    end

    subgraph L1["Ethereum (L1)"]
        UniL1["ExampleUniswapPortal<br/>(Solidity)"]
        PortalA["ExampleTokenPortal A"]
        PortalB["ExampleTokenPortal B"]
        ERC20A["WETH (ERC20)"]
        ERC20B["DAI (ERC20)"]
    end

    User -->|"1. swap_public()"| UniL2
    UniL2 -->|"2. Transfer WETH"| TokenA
    UniL2 -->|"3. Exit to L1"| BridgeA
    UniL2 -->|"4. L2→L1 msg"| UniL1

    BridgeA -->|"5. L2→L1 msg"| PortalA

    UniL1 -->|"6. Consume msg 1"| PortalA
    UniL1 -->|"7. Consume msg 2"| UniL1
    UniL1 -->|"8. Mock swap"| ERC20B
    UniL1 -->|"9. Deposit DAI"| PortalB

    PortalB -->|"10. L1→L2 msg"| BridgeB
    BridgeB -->|"11. Mint DAI"| TokenB
    TokenB -->|"12. DAI to user"| User
```

Each swap generates **two L2-to-L1 messages**:

1. **Token bridge exit** - Instructs the token portal to release input tokens to the uniswap portal
2. **Swap intent** - Proves the user authorized this exact swap with these exact parameters

Both messages must be consumed on L1 before the swap executes. This two-message pattern prevents anyone from stealing funds or changing swap parameters.

> ### Why Two Messages?

Reviewer: this section is redundant with the lines just above. consider consolidating them

:::

#include_code example_uniswap_portal /docs/examples/solidity/example_swap/ExampleUniswapPortal.sol solidity

The public swap function consumes two messages and deposits the output:

#include_code swap_public /docs/examples/solidity/example_swap/ExampleUniswapPortal.sol solidity

The private swap follows the same pattern but deposits output tokens privately:

#include_code swap_private /docs/examples/solidity/example_swap/ExampleUniswapPortal.sol solidity

## Part 3: Uniswap Contract (Noir)

The L2 contract handles the user-facing logic: transferring input tokens, calling the bridge to exit to L1, and creating the swap intent message.

### Setup

The contract stores the portal address and imports the `Token` and `TokenBridge` contracts:

#include_code example_uniswap_setup /docs/examples/contracts/example_uniswap/src/main.nr rust

### Public Swap

The public swap transfers tokens from the sender to the contract, exits them to L1 via the bridge, and sends a swap intent message:

:::note Authorization Witnesses

> Aztec uses **authorization witnesses** (authwit) instead of the ERC20 approve/transferFrom pattern. The contract computes the hash of the exact action it wants to perform, sets that hash as authorized, then immediately performs the action. This gives fine-grained control - the authorization is for a specific action, not a blanket approval. Since we authorize and spend in the same transaction, replay attacks are impossible.

Reviewer: maybe add a link to the authwit page

sidebar_position: 2
description: "Build a cross-chain token swap that exits L2, swaps on L1, and deposits the output back to L2 using Aztec's messaging protocol."
references: ["docs/examples/contracts/example_uniswap/*", "docs/examples/solidity/example_swap/*", "docs/examples/ts/example_swap/*"]
---

## Why Build a Cross-Chain Swap?

DeFi liquidity lives on Ethereum L1. Users with tokens on Aztec L2 need a way to access L1 DEXs like Uniswap without manually bridging tokens back and forth. A cross-chain swap automates this: the user initiates the swap on L2, and the protocol handles exiting to L1, performing the swap, and depositing the output back to L2.

This tutorial walks you through building a version of this flow. You will learn how [**L2-to-L1 messages**](../../aztec-nr/framework-description/ethereum_aztec_messaging.md) work and how multiple contracts across two chains coordinate to execute a single user action.

## Prerequisites

Before starting this tutorial, you need:

1. **Aztec local network** running at version #include_aztec_version -- see [the local network guide](../../../getting_started_on_local_network.md) for setup instructions

> 2. **Foundry** installed for Solidity compilation -- see [Foundry installation](https://book.getfoundry.sh/getting-started/installation)

Reviewer: ```suggestion
```
comes with aztec

Reviewer: if you accept this, you'll also have to renumber the following items

3. **Node.js** (v18+) and a package manager (yarn or npm)
4. **Familiarity with the token bridge tutorial** -- this tutorial builds on concepts from [Bridge Your NFT to Aztec](./token_bridge.md), especially portal contracts and cross-chain messaging

:::tip Background Knowledge
If you are new to Aztec's cross-chain architecture, review [Ethereum-Aztec Messaging](../../aztec-nr/framework-description/ethereum_aztec_messaging.md) first. Key concepts used throughout this tutorial:

- **Portals** -- L1 contracts that communicate with L2 contracts via the Aztec messaging protocol
- **L2-to-L1 messages** -- messages sent from Aztec to Ethereum, stored in a Merkle tree and consumed on L1
- **Authorization witnesses (authwit)** -- Aztec's alternative to ERC20 approve/transferFrom ([learn more](../../aztec-nr/framework-description/authentication_witnesses.md))
- **Content hashes** -- cryptographic digests that uniquely identify cross-chain messages, ensuring L1 and L2 agree on message parameters
:::

## Project Setup

> This tutorial uses three types of contracts (Solidity, Noir, TypeScript) spread across a project. The code is shown inline via snippets -- you do not need to type it manually. However, understanding the project layout will help you follow along.

Reviewer: You mention here that readers do not need to type this out manually. But at the end of Step 2 (Compile Solidity Contracts) has an instruction to run `forge build`. At that point, I was just following along, reading, so I don't have a project set up. It's not clear if I should be building a project as follow along or just reading.  Since it's a tutorial, let's have readers build the project as they go, as much as possible, so we need to add some steps about how to set up the project (either clone a template repo, or specify what commands they should run to get the project set up)

- `keccak256` -- for computing Solidity-compatible function selectors

**TypeScript** (in `package.json`):
- `@aztec/aztec.js`, `@aztec/accounts`, `@aztec/wallets`, `@aztec/stdlib`
- `@aztec/ethereum`, `@aztec/noir-contracts.js`, `@aztec/foundation`

:::note
The TypeScript script imports compiled Solidity artifacts and a generated Noir artifact (`ExampleUniswapContract`). You will compile both before running the script.
:::

## What You'll Build

```mermaid
graph TB
    subgraph L2["Aztec (L2)"]
        User["User"]
        UniL2["ExampleUniswap<br/>(Noir)"]
        BridgeA["TokenBridge A"]
        BridgeB["TokenBridge B"]
        TokenA["Token A (WETH)"]
        TokenB["Token B (DAI)"]
    end

    subgraph L1["Ethereum (L1)"]
        UniL1["ExampleUniswapPortal<br/>(Solidity)"]
        PortalA["ExampleTokenPortal A"]
        PortalB["ExampleTokenPortal B"]
        ERC20A["WETH (ERC20)"]
        ERC20B["DAI (ERC20)"]
    end

    User -->|"1. swap_public()"| UniL2
    UniL2 -->|"2. Transfer WETH"| TokenA
    UniL2 -->|"3. Exit to L1"| BridgeA
    UniL2 -->|"4. L2→L1 msg"| UniL1

    BridgeA -->|"5. L2→L1 msg"| PortalA

    UniL1 -->|"6. Consume msg 1"| PortalA

>     UniL1 -->|"7. Consume msg 2"| UniL1

Reviewer: Should this arrow point to the outbox? Its confusing that that the UniL1 box is pointing to itself

    UniL1 -->|"8. Mock swap"| ERC20B
    UniL1 -->|"9. Deposit DAI"| PortalB

    PortalB -->|"10. L1→L2 msg"| BridgeB
    BridgeB -->|"11. Mint DAI"| TokenB
    TokenB -->|"12. DAI to user"| User
```

Each swap generates **two L2-to-L1 messages**, both of which must be consumed on L1 before the swap executes:

> 1. **Token bridge exit** - Authorizes releasing input tokens from the bridge to the uniswap portal

Reviewer: ```suggestion
1. **Token bridge exit** - Authorizes releasing input tokens from the token portal to the uniswap portal
```
Since its referred to as TokenPortal.sol in the contract name above

2. **Swap intent** - Proves the user authorized *this specific swap* with *these exact parameters*

Neither message alone is sufficient. If only the token exit existed, anyone observing it could potentially redirect the swap. If only the swap intent existed, there would be no proof that tokens were actually withdrawn. Together, they create a cryptographic chain of authorization. This two-message pattern is common in Aztec cross-chain applications where multiple independent systems must coordinate.

## Part 1: Token Portal (Solidity)

> The [token portal](../../aztec-nr/framework-description/ethereum_aztec_messaging.md) handles depositing tokens from L1 to L2 and withdrawing from L2 to L1. This version hardcodes `caller_on_l1` to `address(0)` (meaning anyone can execute the withdrawal).

Reviewer: I dont see caller_on_l1 being hardcoded to 0. maybe i missing it?

Reviewer: also, I think we can link to a better page, the current link is about general message passing, not about a token portal specifically, as the link indicates. maybe a linke to the aztecjs token bridge tutorial would be better?

#include_code example_token_portal /docs/examples/solidity/example_swap/ExampleTokenPortal.sol solidity

Key functions:

- `depositToAztecPublic` - Locks ERC20 tokens and sends an L1→L2 message for public minting
- `depositToAztecPrivate` - Same but for private minting
- `withdraw` - Consumes an L2→L1 message and releases tokens

The [registry](../../aztec-nr/framework-description/ethereum_aztec_messaging.md) provides governance-updateable addresses for core Aztec contracts. Rather than hardcoding rollup addresses, portals query the registry, allowing the protocol to upgrade without redeploying all portals.

> Each cross-chain message includes a **content hash** — a `keccak256` digest of the function selector and its parameters that uniquely identifies the message. The content hash in each function uses `abi.encodeWithSignature` to include a function selector, making each message type unique and preventing a deposit message from being confused with a withdrawal message.

Reviewer: isnt it sha256, not keccak?

## Part 3: Uniswap Contract (Noir)

The L2 contract handles the user-facing logic: transferring input tokens, calling the bridge to exit to L1, and creating the swap intent message.

### Setup

The contract stores the portal address and imports the `Token` and `TokenBridge` contracts:

#include_code example_uniswap_setup /docs/examples/contracts/example_uniswap/src/main.nr rust

### Public Swap

The public swap transfers tokens from the sender to the contract, exits them to L1 via the bridge, and sends a swap intent message:

:::note Authorization Witnesses
Aztec uses [**authorization witnesses** (authwit)](../../aztec-nr/framework-description/authentication_witnesses.md) instead of the ERC20 approve/transferFrom pattern. The contract computes the hash of the exact action it wants to perform, sets that hash as authorized, then immediately performs the action. This gives fine-grained control - the authorization is for a specific action, not a blanket approval. Since we authorize and spend in the same transaction, replay attacks are impossible.
:::

#include_code swap_public /docs/examples/contracts/example_uniswap/src/main.nr rust

### Private Swap

> The private swap is similar but uses `transfer_to_public` (private to public transfer) and [`enqueue_self`](../../aztec-nr/framework-description/calling_contracts.md) instead of `call_self`:

Reviewer: I think a brief inline explanation of why enqueue_self instead of call_self would be helpful

#include_code swap_private /docs/examples/contracts/example_uniswap/src/main.nr rust

:::note Why no recipient parameter?

> In `swap_private`, the recipient is the transaction sender (implicit). This preserves privacy: revealing a recipient address to L1 would compromise the caller's identity. The output tokens are deposited privately on L2, where only the caller can claim them.

Reviewer: ```suggestion
In `swap_private`, the recipient is the person that provide the secret used to generate the hash for the L1 to L2 message. This preserves privacy: revealing a recipient address to L1 would compromise the caller's identity. The output tokens are deposited privately to L2, where only the secret holder can claim them.
```
I think this is more accurate

First, read the rollup version from the portal and compute the content hash and message leaf for the token bridge exit message:

#include_code consume_l1_messages_setup /docs/examples/ts/example_swap/index.ts typescript

Next, compute Merkle membership witnesses for both L2→L1 messages. The outbox stores messages in a Merkle tree, so we need the sibling path (proof of inclusion) for each message. We also compute the swap intent message leaf using the same encoding as the Solidity portal:

#include_code consume_l1_messages_witnesses /docs/examples/ts/example_swap/index.ts typescript

Finally, call `swapPublic` on the L1 uniswap portal, passing both message proofs. The portal verifies both messages against the outbox, performs the mock swap, and deposits the output tokens back to L2:

#include_code consume_l1_messages_execute /docs/examples/ts/example_swap/index.ts typescript

> Finally, claim the output DAI on L2:

Reviewer: You also start line 306 with `Finally,`

sidebar_position: 2
description: "Build a cross-chain token swap that exits L2, swaps on L1, and deposits the output back to L2 using Aztec's messaging protocol."
references: ["docs/examples/contracts/example_uniswap/*", "docs/examples/solidity/example_swap/*", "docs/examples/ts/example_swap/*"]
---

## Why Build a Cross-Chain Swap?

DeFi liquidity lives on Ethereum L1. Users with tokens on Aztec L2 need a way to access L1 DEXs like Uniswap without manually bridging tokens back and forth. A cross-chain swap automates this: the user initiates the swap on L2, and the protocol handles exiting to L1, performing the swap, and depositing the output back to L2.

This tutorial walks you through building a version of this flow. You will learn how [**L2-to-L1 messages**](../../aztec-nr/framework-description/ethereum_aztec_messaging.md) work and how multiple contracts across two chains coordinate to execute a single user action.

## Prerequisites

Before starting this tutorial, you need:

1. **Aztec local network** running at version #include_aztec_version -- see [the local network guide](../../../getting_started_on_local_network.md) for setup instructions

> 2. **Node.js** (v18+) and a package manager (yarn or npm)

Reviewer: ```suggestion
2. **Node.js** (v24+) and a package manager (yarn or npm)
```
I think we are recommending node 24 now? Its what we use in aztec-packages

:::tip Background Knowledge
If you are new to Aztec's cross-chain architecture, review [Ethereum-Aztec Messaging](../../aztec-nr/framework-description/ethereum_aztec_messaging.md) first. Key concepts used throughout this tutorial:

- **Portals** -- L1 contracts that communicate with L2 contracts via the Aztec messaging protocol
- **L2-to-L1 messages** -- messages sent from Aztec to Ethereum, stored in a Merkle tree and consumed on L1
- **Authorization witnesses (authwit)** -- Aztec's alternative to ERC20 approve/transferFrom ([learn more](../../aztec-nr/framework-description/authentication_witnesses.md))
- **Content hashes** -- cryptographic digests that uniquely identify cross-chain messages, ensuring L1 and L2 agree on message parameters
:::

## Project Setup

This tutorial walks you through three types of contracts (Solidity, Noir, TypeScript) that work together. You will clone the example project and build it as you follow along.

### Clone the Example Code

The example code lives in the Aztec packages repository:

```bash

> git clone --depth 1 --branch #include_aztec_version https://github.com/AztecProtocol/aztec-packages

Reviewer: ```suggestion
git clone --depth 1 --branch #include_aztec_version https://github.com/AztecProtocol/aztec-packages.git
```

cd aztec-packages/docs/examples
```

### Project Structure

The relevant files are spread across three directories:

```
examples/

> ├── solidity/

Reviewer: nit: this actually comes after contracts/

```

### Dependencies

**Solidity** (via Foundry import mappings in `foundry.toml`):
- OpenZeppelin ERC20 (`@oz/token/ERC20/`)
- Aztec L1 contracts (`@aztec/core`, `@aztec/governance`)

**Noir** (in `Nargo.toml`):
- `aztec` -- the Aztec Noir framework
- `token` -- the standard Token contract
- `token_bridge` -- the standard TokenBridge contract
- `keccak256` -- for computing Solidity-compatible function selectors

**TypeScript**:
- `@aztec/aztec.js`, `@aztec/accounts`, `@aztec/wallets`, `@aztec/stdlib`
- `@aztec/ethereum`, `@aztec/noir-contracts.js`, `@aztec/foundation`

:::note
The TypeScript script imports compiled Solidity artifacts and a generated Noir artifact (`ExampleUniswapContract`). You will compile both before running the script.
:::

## What You'll Build

```mermaid
graph TB
    subgraph L2["Aztec (L2)"]
        User["User"]
        UniL2["ExampleUniswap<br/>(Noir)"]
        BridgeA["TokenBridge A"]
        BridgeB["TokenBridge B"]
        TokenA["Token A (WETH)"]
        TokenB["Token B (DAI)"]
    end

    subgraph L1["Ethereum (L1)"]
        UniL1["ExampleUniswapPortal<br/>(Solidity)"]
        PortalA["ExampleTokenPortal A"]
        PortalB["ExampleTokenPortal B"]

>         ERC20A["WETH (ERC20)"]

Reviewer: this one isnt used in the diagram flow. do we need it? 

<img width="768" height="344" alt="Image" src="https://github.com/user-attachments/assets/2d87daf8-bc2c-4c32-9d81-b8fc90099215" />

        ERC20B["DAI (ERC20)"]
        Outbox["Outbox"]
    end

    User -->|"1. swap_public()"| UniL2
    UniL2 -->|"2. Transfer WETH"| TokenA
    UniL2 -->|"3. Exit to L1"| BridgeA

>     UniL2 -->|"4. L2→L1 msg"| UniL1

Reviewer: its kind of confusing that the numbers don't flow from one box to the next, they kind of hop around. adding a return arrow might help (e.g. for 2 and 3). Also 4 and 5 have the same message "L2->L1 msg" which is confusing

<img width="862" height="578" alt="Image" src="https://github.com/user-attachments/assets/dd1ffdbb-9147-4e6a-95b3-b32b7f352502" />

### Compile and Generate Bindings

From the `examples/contracts/example_uniswap` directory, compile the Noir contract and generate TypeScript bindings:

```bash
cd ../contracts/example_uniswap
aztec compile
aztec codegen target -o ../../ts/example_swap/artifacts
```

The `aztec compile` command compiles the Noir contract. The `aztec codegen` command generates a TypeScript class (`ExampleUniswapContract`) from the compiled artifact, which you will use in the deployment script.

:::note
Before proceeding, make sure you have compiled both the Solidity contracts (Part 1-2) and the Noir contract (Part 3). The TypeScript script below imports compiled artifacts from both.
:::

## Install TypeScript Dependencies

From the `examples/ts/example_swap` directory, initialize a project and install the required packages:

```bash
cd ../../ts/example_swap
npm init -y
npm install \
  @aztec/aztec.js@#include_aztec_version \
  @aztec/accounts@#include_aztec_version \
  @aztec/wallets@#include_aztec_version \
  @aztec/stdlib@#include_aztec_version \
  @aztec/ethereum@#include_aztec_version \
  @aztec/noir-contracts.js@#include_aztec_version \

>   @aztec/foundation@#include_aztec_version \

Reviewer: looks like viem also required? its in the config.yaml
